### DISCUSSION TITLE
shell-script mode company-backends

### DISCUSSION THREAD

**Discussion by ville-h** [company-mode/company-mode#1383] [Q&A] [answered]
Out-of-the-box company-mode(?) doesn't seem very usable when writing bash scripts. It seems to take ~15 seconds for the tooltip to show and whilst doing so locking up emacs. Is this the expected behaviour? It would appear my `company-backends` is:

```
(company-bbdb company-semantic company-cmake company-capf company-clang company-files
    (company-dabbrev-code company-gtags company-etags company-keywords)
    company-oddmuse company-dabbrev)
```

**Comment by dgutov** [comment]
Which backend takes ~15 seconds? What does `M-x company-diag` say?

**Comment by ville-h** [comment]
I wouldn't know how to tell which backend is taking how much time. As far as `company-diag` do you want me to paste its output? It's largely big list of the completions, but here's the beginning of it:
```
Used backend: company-capf
Value of c-a-p-f: (sh-completion-at-point-function comint-completion-at-point t)
Major mode: sh-mode
Prefix: "r"
Completions: ...
```

On further testing large part of the ~15 seconds is caused by enabling prescient. So perhaps they should be asked about their performance. With out prescient it's still ~3 seconds.

**Reply by dgutov** [reply]
> On further testing large part of the ~15 seconds is caused by enabling prescient. So perhaps they should be asked about their performance.

That's definitely a good idea.

> With out prescient it's still ~3 seconds.

You could try finding out whether it's the completion from `sh-completion-at-point-function`, or the next one.

Also try `C-M-i`: if completion takes 3 seconds with it as well, it will be a good occasion to write a report with `M-x report-emacs-bug`.

**Comment by ville-h** [comment]
> You could try finding out whether it's the completion from sh-completion-at-point-function, or the next one.`

I do not know how to do that as I don't know how to program in lisp. What is "the next one" in this context?

> Also try C-M-i

I tried calling `completion-at-point` directly and it was near-instant. Guessing it was in the <150ms where I couldn't tell that there was a lag between me hitting enter to call it and result showing up.

**Reply by dgutov** [reply]
> I do not know how to do that as I don't know how to program in lisp. What is "the next one" in this context?

The "next one" is `comint-completion-at-point` in this context. But measuring them without some skill in Lisp debugging might indeed be non-trivial.

> I tried calling completion-at-point directly and it was near-instant. Guessing it was in the <150ms where I couldn't tell that there was a lag between me hitting enter to call it and result showing up.

Okay. And you see the same completions as the ones that you see with Company (delayed by 3 seconds)?

**Comment by ville-h** [comment]
The lists are fairly long so I didn't check if they match fully, but the handful of candidates I picked from one list was also in the other list.

**Reply by dgutov** [reply]
Okay. Let's try the profiler.

Start it with `M-x profiler-start`, choose `cpu`.

Then perform a few completions (ones that take ~3 seconds in your estimation). With `prescient` disabled, of course.

Then `M-x profiler-report`. You can expand the "interesting" nodes (ones with larger numbers in the `%` column) in the output with TAB and copy and paste the result here.

**Comment by ville-h** [comment, marked as answer]
It would seem like this was some kind of emacs28 bug. I've since upgraded to emacs29 and even with prescient enabled the performance is in the "interactive-ish" region of <1.0 seconds. [prescient discussion](https://github.com/radian-software/prescient.el/issues/144).
